### PR TITLE
OSAC 89 - Bulk Insert activity

### PR DESCRIPTION
**Pull request by alcxs**
Description
New activity for bulk insert into database table.
How to test
add the activity in the designer and test functionality
Jira
[OSAC-89](https://uipath.atlassian.net/browse/OSAC-89)
Label:
24S

### CODE REVIEW

File: Activities/Database/UiPath.Database/DatabaseConnection.cs [hunk 1]
```diff
@@ -94,6 +99,182 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
...
+            dbDA.ContinueUpdateOnError = false;
+
+            long affectedRecords = 0;
+
+           
+
+            if (_providerName == "System.Data.SqlClient")
+            {
+                //if SQL Server
+
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database/DatabaseConnection.cs]: we could extract code in separate methods. this body is pretty big

File: Activities/Database/UiPath.Database/DatabaseConnection.cs [hunk 2]
```diff
@@ -94,6 +99,182 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
...
+                using (SqlBulkCopy bulkCopy = new SqlBulkCopy(_connection.ConnectionString, SqlBulkCopyOptions.UseInternalTransaction))
+                {
+                    bulkCopy.DestinationTableName = tableName;
+                    bulkCopy.WriteToServer(dataTable);
+                }
+
+                // Perform a final count on the destination
+                // table to see how many rows were added.
+                long countEnd = Convert.ToInt32(
+              commandRowCount.ExecuteScalar());
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database/DatabaseConnection.cs]: move to above line?

File: Activities/Database/UiPath.Database/DatabaseConnection.cs [hunk 3]
```diff
@@ -94,6 +99,182 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
...
+                var ds = new DataSet();
+                dbDA.Fill(ds);
+                ValidateTableStructure(tableName, dataTable, ds);
+
+                // Perform an initial count on the destination table.
+                SqlCommand commandRowCount = new SqlCommand(string.Format("SELECT COUNT(*) FROM {0}", tableName), (SqlConnection)_connection);
+                long countStart = Convert.ToInt32(
+                    commandRowCount.ExecuteScalar());
+                // Set up the bulk copy object
+                using (SqlBulkCopy bulkCopy = new SqlBulkCopy(_connection.ConnectionString, SqlBulkCopyOptions.UseInternalTransaction))
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database/DatabaseConnection.cs]: not sure if we can moq the SqlBulkCopy object directly but if we wrap this in a factory method and return an interface we could write some tests for the rest of the code

File: Activities/Database/UiPath.Database/DatabaseConnection.cs [hunk 4]
```diff
@@ -94,6 +99,182 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
...
+                    //try all drivers with latest version first
+                    foreach (var item in oracleList.OrderByDescending(x => x))
+                    {
+
+                        var oracle = Assembly.LoadFile(item);
+                        var bulkCopyType = oracle.GetType("Oracle.ManagedDataAccess.Client.OracleBulkCopy");
+
+                        if (bulkCopyType != null)
+                        {
+                            dynamic bulkCopy = Activator.CreateInstance(bulkCopyType, new object[] { connection });
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database/DatabaseConnection.cs]: same here. if make some utility interfaces we could test most of the logic. all we need to abstract is the call that goes to the db (WriteToServer in this case)

**Review by adrian-zanescu — CHANGES_REQUESTED** [review]

**Comment by adrian-zanescu** [comment]
For warnings using the robot api use https://www.myget.org/feed/workflow/package/nuget/UiPath.Activities.Api/20.4.0 as a dependency in the project. Then in the execute code of the activity do:

`var runtime = context.GetExtension<IExecutorRuntime>();
if(runtime.HasFeature(ExecutorFeatureKeys.LogMessage))
{
    runtime.LogMessage(..................);
}}`

File: Activities/Database/UiPath.Database/DatabaseConnection.cs (hunk 1 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database/DatabaseConnection.cs]: method extracted separately

File: Activities/Database/UiPath.Database/DatabaseConnection.cs (hunk 2 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database/DatabaseConnection.cs]: moved

File: Activities/Database/UiPath.Database/DatabaseConnection.cs (hunk 3 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database/DatabaseConnection.cs]: extracted to interface in a separate file.

File: Activities/Database/UiPath.Database/DatabaseConnection.cs (hunk 4 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database/DatabaseConnection.cs]: extracted to interface

File: Activities/Database/UiPath.Database/UiPath.Database.csproj [hunk 5]
```diff
@@ -22,6 +23,10 @@
     <PackageReference Include="UiPath.Workflow" />
     <PackageReference Include="System.Activities.Core.Presentation" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Data.SqlClient" Version="4.8.2" />
+    <PackageReference Include="UiPath.Activities.Api" />
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database/UiPath.Database.csproj]: Make it Private>All - this is just a development dependency. The robot will provide the actual dll at runtime

File: Activities/Database/UiPath.Database/DatabaseConnection.cs [hunk 6]
```diff
@@ -94,6 +101,214 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
...
+                    }
+                    else continue;
+                }
+
+                if (!bulkCopyIsPresent)
+                {
+                    //if no bulk insert possible, fallback to insert data table
+                    if (executorRuntime != null)
+                    {
+                        var message = new LogMessage
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database/DatabaseConnection.cs]: extract in a separate function - actually 2 of the. first one checks executorRuntime.HasFeature( and calls the second one that performs the logging

File: Activities/Database/UiPath.Database.Activities.Packaging/UiPath.Database.Activities.Packaging.csproj [hunk 7]
```diff
@@ -31,9 +32,9 @@
 		<ItemGroup>
 			<BuildOutputInPackage Include="$(OutputPath)UiPath.Database.dll" />
 			<BuildOutputInPackage Include="$(OutputPath)UiPath.Database.Activities.dll" />
-			<BuildOutputInPackage Include="$(OutputPath)Microsoft.Data.ConnectionUI.Dialog.dll" />
-
-			<BuildOutputInPackage Include="$(OutputPath)fr\UiPath.Database.resources.dll" TargetPath="fr\UiPath.Database.resources.dll" />
+      <BuildOutputInPackage Include="$(OutputPath)Microsoft.Data.ConnectionUI.Dialog.dll" />
```
Comment (adrian-zanescu) [inline Activities/Database/UiPath.Database.Activities.Packaging/UiPath.Database.Activities.Packaging.csproj:35]: spacing

**Review by adrian-zanescu — CHANGES_REQUESTED** [review]
still no unit tests. we could have at least some coverage

File: Activities/Database/UiPath.Database/UiPath.Database.csproj (hunk 5 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database/UiPath.Database.csproj]: made private, as requested

File: Activities/Database/UiPath.Database/DatabaseConnection.cs (hunk 6 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database/DatabaseConnection.cs]: already separate method

File: Activities/Database/UiPath.Database.Activities.Packaging/UiPath.Database.Activities.Packaging.csproj (hunk 7 above)
Comment (alcxs) [reply Activities/Database/UiPath.Database.Activities.Packaging/UiPath.Database.Activities.Packaging.csproj:35]: cleanup done on all csproj files

**Comment by alcxs** [comment]
> still no unit tests. we could have at least some coverage

added unit test to verify logic for bulk insert

**Review by adrian-zanescu — APPROVED** [review]

**Pull request merged**